bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/handler/EcovacsApiHandler.java

```diff
@@ -0,0 +1,173 @@
+    }
+
+    @Nullable
+    public EcovacsApi createApiForDevice(String serial) {
+        String country = localeProvider.getLocale().getCountry();
+        EcovacsApi api = this.api;
+        if (api == null || country.isEmpty()) {
+            return null;
+        }
+        return createApi("-" + serial, country);
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing Ecovacs account '{}'", getThing().getUID().getId());
+        // The API expects us to provide a unique device ID during authentication, so generate one once
+        // and keep it in configuration afterwards
+        if (!getConfig().keySet().contains("installId")) {
+            Configuration newConfig = editConfiguration();
+            newConfig.put("installId", UUID.randomUUID().toString());
+            updateConfiguration(newConfig);
+        }
+        initializeApi();
+    }
+
+    @Override
+    public void dispose() {
+        super.dispose();
+        final EcovacsDeviceDiscoveryService discoveryService = this.discoveryService;
+        if (discoveryService != null) {
+            discoveryService.stopScan();
+        }
+    }
+
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Collections.singleton(EcovacsDeviceDiscoveryService.class);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (RefreshType.REFRESH == command) {
+            logger.debug("Refreshing Ecovacs API account '{}'", getThing().getUID().getId());
+            initializeApi();
+        }
+    }
+
+    public void onLoginExpired() {
+        logger.debug("Ecovacs API login for account '{}' expired, logging in again", getThing().getUID().getId());
+        final EcovacsApi api = this.api;
+        if (api != null) {
+            loginToApi(api);
+        }
+    }
+
+    private void initializeApi() {
+        String country = localeProvider.getLocale().getCountry();
+        if (country.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "@text/offline.config-error-no-country");
+            return;
+        }
+
+        EcovacsApi api = createApi("", country);
+        loginToApi(api);
+    }
+
+    private EcovacsApi createApi(String deviceIdSuffix, String country) {
+        EcovacsApiConfiguration config = getConfigAs(EcovacsApiConfiguration.class);
+        String deviceId = config.installId + deviceIdSuffix;
+        org.openhab.binding.ecovacs.internal.api.EcovacsApiConfiguration apiConfig = new org.openhab.binding.ecovacs.internal.api.EcovacsApiConfiguration(
+                deviceId, config.email, config.password, config.continent, country, "EN", ClientKeys.CLIENT_KEY,
+                ClientKeys.CLIENT_SECRET, ClientKeys.AUTH_CLIENT_KEY, ClientKeys.AUTH_CLIENT_SECRET);
+
+        return EcovacsApi.create(httpClientFactory.getCommonHttpClient(), apiConfig);
```

Comment:
If only common client is used, you could consider refactoring the constructor to simply take `HttpClient` rather than `HttpClientFactory`.